File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> prev_chars;
        int w_size = 0;
        int max_len = 0;

        for(int i = 0; i < s.size(); i++){
            if(i > 0){
                prev_chars.erase(s[i-1]);
                w_size--;
            }

            while(i+w_size < s.size() && prev_chars.count(s[i+w_size]) == 0){
                prev_chars.insert(s[i+w_size]);
                w_size++;
            }
            max_len = max(max_len, w_size);
        }
        return max_len;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
from collections import defaultdict
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # substring -> sliding window
        # how to check if there is a duplicate char -> dict -> use set to store the element in the current window 
        # shrink left until the set does not contain the rightmost element
        result_len = 0
        chars_in_window = defaultdict(int)
        left = 0

        for right in range(len(s)):
            chars_in_window[s[right]] += 1

            while left < right and chars_in_window[s[right]] > 1:
                chars_in_window[s[left]] -= 1
                left += 1

            result_len = max(result_len, right - left + 1)
        
        return result_len

     